pkg/nodes/list.go

```diff
@@ -60,3 +68,115 @@ func ListExternalIPv4Networks(apiClient *clients.Settings, options v1.ListOption
 
 	return ipV4ExternalAddresses, nil
 }
+
+// WaitForAllNodesAreReady waits for all nodes to be Ready for a time duration up to the timeout.
+func WaitForAllNodesAreReady(apiClient *clients.Settings,
+	timeout time.Duration,
+	options v1.ListOptions) (bool, error) {
+	glog.V(100).Infof("Waiting for all nodes to be in the Ready state for up to a duration of %v",
+		timeout)
+
+	nodesList, err := List(apiClient, options)
+	if err != nil {
+		glog.V(100).Infof("Failed to list all nodes due to %s", err.Error())
+
+		return false, err
+	}
+
+	// Wait 5 secs in each iteration before condition function () returns true or errors or times out
+	// after availableDuration
+	err = wait.PollImmediate(backoff, timeout, func() (done bool, err error) {
+		for _, node := range nodesList {
+			ready, err := node.IsReady()
+			if err != nil {
+				glog.V(100).Infof("Node %v has error %w", node.Object.Name, err)
+
+				return false, err
+			}
+
+			if !ready {
+				glog.V(100).Infof("Node %s not Ready", node.Object.Name)
+
+				return false, nil
+			}
+		}
+
+		// Here we are always returning "false, nil" so we keep iterating throughout the stableInterval
+		// of the inner wait.pollImmediate loop, until we time out.
+		return true, nil
+	})
+
+	if err == nil {
+		glog.V(100).Infof("All nodes were found in the Ready State during availableDuration: %v",
+			timeout)
+
+		return true, nil
+	}
+
+	// Here err is "timed out waiting for the condition"
+	glog.V(100).Infof("Not all nodes were found in the Ready State during availableDuration: %v",
+		err)
+
+	return false, err
+}
+
+// WaitForAllNodesToReboot waits for all nodes to start and finish reboot up to the timeout.
+func WaitForAllNodesToReboot(apiClient *clients.Settings,
+	options v1.ListOptions,
+	singleNodeChangeStateTimeout time.Duration) (bool, error) {
+	glog.V(100).Infof("Waiting for all nodes in the list to reboot and return to the Ready condition")
+
+	nodesList, err := List(apiClient, options)
+	if err != nil {
+		glog.V(100).Infof("Failed to list all nodes due to %s", err.Error())
+
+		return false, err
+	}
+
+	globalRebootTimeout := 2 * time.Duration(len(nodesList)) * singleNodeChangeStateTimeout
```

Comment:
I think what you did here makes sense but I'm concerned that objects may change state between iterations (say a node times out waiting to be not ready and then on the next iteration it is ready. This node would be orphaned). WDYT about tracking the state of rebooted nodes similarly to the way you are tracking ready nodes? Something like this:
```
	globalStartTime := time.Now().Unix()
	readyNodes := []string{}
	rebootedNodes := []string{}
	err = wait.PollImmediate(backoff, globalRebootTimeout, func() (done bool, err error) {
		for _, node := range nodesList {
			if !slices.Contains(readyNodes, node.Object.Name) {
				ready, err := node.IsReady()
				if err != nil {
					return false, err
				}

				rebooted := slices.Contains(rebootedNodes, node.Object.Name)
				if !ready && !rebooted {
					glog.V(100).Infof("Node %s was rebooted and is starting to recover", node.Object.Name)
					
					rebootedNodes = append(rebootedNodes, node.Object.Name)
				}

				if ready && rebooted {
					glog.V(100).Infof("Node %s was successfully rebooted after: %v", time.Now().Unix() - globalStartTime)
					
					readyNodes = append(readyNodes, node.Object.Name)
				}
			}
		}

		return len(readyNodes) == len(nodesList), nil
```
I think you can also just use the timeout provided by the user for this. That way the timeout is predictable and we don't have to calculate our own timeout.

Reply:
testing required, I will confirm when I'll have any result

Reply:
tested